### ISSUE TITLE
Client response-parsing very slow

### ISSUE THREAD

**Issue by MatthewSteel** [googleapis/nodejs-bigquery#1232]
When the response is more than a couple hundred kilobytes the nodejs client library adds substantial overhead, resulting in total running time several times (5x?) higher than is necessary.

I took a quick profile and I _think_ (60% sure) the time is predominantly being spent in,
* [This call](https://github.com/googleapis/nodejs-common/blob/89890df3d8e7fb0b48548b85f4b46188ef4993c4/src/util.ts#L399) to `extend`, and
* [This merge-schema stuff](https://github.com/googleapis/nodejs-bigquery/blob/51fc8c37f3c4211bee85de4f094e9a2e25cd898b/src/job.ts#L562).

In my testing I get reliable, substantial speedups from rewriting my queries from

```sql
select * from tbl order by bar
```
to
```sql
with d as (
  -- most of the original query
  select * from tbl
) select to_json_string((select array(
                            -- original "order by"
  select as struct * from d order by bar
)))
```
and then immediately calling `JSON.parse` on the first row.

Sorry for my bad SQL, and sorry if this has been brought up already. I guess returning "biggish" responses is a bit unusual, but I bet it affects lots of people, and I think it adds a lot of overhead even when people use pagination.

I think this issue is related to [this stale stackoverflow question](https://stackoverflow.com/questions/49283553/bigquery-node-api-is-much-slower-than-the-gui), and probably also [this one](https://stackoverflow.com/questions/56433464/google-bigquery-python-library-is-2x-as-fast-as-the-node-js-library-for-download).

#### Environment details

  - OS: linux,
  - Node.js version: 16.17
  - npm version: ?
  - `@google-cloud/bigquery` version: 6.2.1

#### Steps to reproduce

I'd _guess_ this is reproducible with the client library by running something like

```sql
select generate_array(1, 100) as foo from unnest(generate_array(1, 10000));
```
with and without the CTE/JSON wrapper, inside the bigquery UI and using the nodejs client.

**Comment by alvarowolfx** [comment]
hey @MatthewSteel thanks for the report. I was looking into that and seeing if some performance improvements can be made in regards to message parsing. Unfortunately what I found ( and kind of expected ) is that most of the time is spent downloading the data itself. Parsing the data takes around 5% of the whole process. In Python and Go SDK, we have some integrations to use [Storage Read API](https://cloud.google.com/bigquery/docs/reference/storage) to fetch data faster.

Here are some results for some queries and using `pprof` to analyze performance:

Query: `select generate_array(1, 100) as foo from unnest(generate_array(1, 10000))` 
![26713-random-10000](https://github.com/googleapis/nodejs-bigquery/assets/1615543/8174ff6d-1f2b-464a-97ae-9cc0c00a111e)

Query: `select generate_array(1, 100) as foo from unnest(generate_array(1, 100000))` 
![29716-random-100000](https://github.com/googleapis/nodejs-bigquery/assets/1615543/b9f0c1db-e0e8-4fa6-ad9a-577d8ac96c6b)

Query: `SELECT * FROM `nyc-tlc.yellow.trips` LIMIT 10000` 
![30294-nyc-10000](https://github.com/googleapis/nodejs-bigquery/assets/1615543/a912f5e3-4a50-4b9c-aba2-932a2e8a6e39)

 Query: ```with d as (
          SELECT * FROM \`nyc-tlc.yellow.trips\` LIMIT 10000
        )
        select to_json_string(
          (
            select array(
              select as struct * from d
            )
          )
        )```
![30771-nyc-10000-json](https://github.com/googleapis/nodejs-bigquery/assets/1615543/07863893-489a-4395-90a2-1a54e0b087a9)

When converting data to JSON using the CTE from BigQuery, there is less usage of functions to convert data, but the download takes almost the same amount of time.

**Comment by MatthewSteel** [comment]
Hmm, this very much does not line up with my experience. My end-to-end latency went down by a factor of about five when turning all the rows into a single JSON blob. The overhead was practically all CPU.

Maybe I have an old version of a transitive dependency -- I made sure to update to the latest bigquery client library before filing the bug, but maybe I should have updated other things too.

**Comment by meredithslota** [comment]
@MatthewSteel — please let us know if you find a bug (with your repro) in the current library + dependencies, and we'll look into it. Otherwise, I'm going to close this out with the general commendation to look into using the BigQuery Storage API for performance-sensitive work.

**Comment by jeff-bradley** [comment]
I also noticed a pretty considerable quicker response when using the to_json_string method. My queries in the BQ UI would be 1 second for returning ~40k rows, and in NodeJS **without** to_json_string it would average around 8 seconds (even before I saw this issue I thought there had to be something up with the conversion). And after changing the SQL to use to_json_string and JSON.parse() in JS the queries are around 3-4 seconds which is a lot better.
